libsweep/src/win/serial.cc

```diff
@@ -0,0 +1,351 @@
+  dcb_serial_params.fDtrControl = DTR_CONTROL_DISABLE;
+
+  // set the serial port parameters to the specified values
+  if (!SetCommState(h_comm, &dcb_serial_params)) {
+    *error = error_construct("setting serial port parameters failed");
+    CloseHandle(h_comm);
+    return nullptr;
+  }
+
+  // specify timeouts (all values in milliseconds)
+  COMMTIMEOUTS timeouts;
+  if (!GetCommTimeouts(h_comm, &timeouts)) {
+    *error = error_construct("retrieving current serial port timeouts failed");
+    CloseHandle(h_comm);
+    return nullptr;
+  }
+  timeouts.ReadIntervalTimeout = 50;         // max time between arrival of two bytes before ReadFile() returns
+  timeouts.ReadTotalTimeoutConstant = 50;    // used to calculate total time-out period for read operations
+  timeouts.ReadTotalTimeoutMultiplier = 10;  // used to calculate total time-out period for read operations
+  timeouts.WriteTotalTimeoutConstant = 50;   // used to calculate total time-out period for write operations
+  timeouts.WriteTotalTimeoutMultiplier = 10; // used to calculate total time-out period for write operations
+
+  // set the timeouts
+  if (!SetCommTimeouts(h_comm, &timeouts)) {
+    *error = error_construct("setting serial port timeouts failed");
+    CloseHandle(h_comm);
+    return nullptr;
+  }
+
+  // create the overlapped event handle for reading
+  OVERLAPPED os_reader = {0};
+
+  // Create the overlapped read event. Must be closed before exiting
+  os_reader.hEvent = CreateEvent(NULL, TRUE, FALSE, NULL);
+  if (os_reader.hEvent == NULL) {
+    *error = error_construct("creating overlapped read event failed");
+    CloseHandle(h_comm);
+    return nullptr;
+  }
+
+  // set the comm mask
+  if (!SetCommMask(h_comm, EV_RXCHAR | EV_ERR)) {
+    *error = error_construct("setting comm mask failed");
+    CloseHandle(h_comm);
+    CloseHandle(os_reader.hEvent);
+    return nullptr;
+  }
+
+  // purge the comm port of any pre-existing data or errors
+  if (!PurgeComm(h_comm, PURGE_RXABORT | PURGE_TXABORT | PURGE_RXCLEAR | PURGE_TXCLEAR)) {
+    *error = error_construct("flushing serial port failed during serial device construction");
+    CloseHandle(h_comm);
+    CloseHandle(os_reader.hEvent);
+    return nullptr;
+  }
+
+  // create the serial device
+  auto out = new device{h_comm, os_reader, FALSE, 500};
+
+  return out;
+}
+
+void device_destruct(device_s serial) {
+  SWEEP_ASSERT(serial);
+
+  error_s ignore = nullptr;
+  device_flush(serial, &ignore);
+
+  // close the serial port
+  CloseHandle(serial->h_comm);
+
+  // close the overlapped read event
+  CloseHandle(serial->os_reader.hEvent);
+
+  (void)ignore; // nothing we can do here
+
+  delete serial;
+}
+
+void device_read(device_s serial, void* to, int32_t len, error_s* error) {
+  SWEEP_ASSERT(serial);
+  SWEEP_ASSERT(to);
+  SWEEP_ASSERT(len >= 0);
+  SWEEP_ASSERT(error);
+
+  DWORD dw_num_read;
+  DWORD dw_num_to_read = (DWORD)len;
+  BOOL f_result = FALSE;
```

Comment:
Is there a reason for BOOL and TRUE / FALSE? What about C++'s bool and true / false?